File: pkg/api/pod.go
package api

import (
	"github.com/spiceai/spiceai/pkg/pods"
)

type Pod struct {
	Name         string   `json:"name,omitempty" csv:"name"`
	ManifestPath string   `json:"manifest_path,omitempty" csv:"manifest_path"`
	Episodes     int      `json:"episodes,omitempty" csv:"episodes"`
	Identifiers  []string `json:"identifiers,omitempty" csv:"-"`
	Measurements []string `json:"measurements,omitempty" csv:"-"`
	Categories   []string `json:"categories,omitempty" csv:"-"`
}

func NewPod(f *pods.Pod) *Pod {
	return &Pod{
		Name:         f.Name,
		Episodes:     f.Episodes(),
		ManifestPath: f.ManifestPath(),
		Identifiers:  f.IdentifierNames(),
		Measurements: f.MeasurementNames(),
		Categories:   f.CategoryNames(),
	}
}

File: dashboard/src/models/pod.ts
import { useState, useEffect } from 'react'
import useSWR from 'swr'

import { getApiPath } from '../util/api'

export interface Pod {
  name: string
  manifest_path: string
  episodes: number
  identifiers: string[]
  measurements: string[]
  categories: string[]
}

export interface PodsResponse {
  data: Pod[]
  isLoading: boolean
  error: any
}

export interface PodResponse {
  data: Pod | undefined
  isLoading: boolean
  error: any
}

const requestInit = {
  headers: new Headers({
    Accept: 'application/json',
  }),
}

const fetcher = async (url: string): Promise<any> => {
  const response = await fetch(url, requestInit)
  return response.json()
}

export function usePods(): PodsResponse {
  const path = getApiPath('/pods')
  const { data, error } = useSWR(path, fetcher, {
    refreshInterval: 250,
  })
  const [pods, setPods] = useState<Pod[]>([])

  useEffect(() => {
    if (data) {
      setPods(data)
    }
  }, [data])

  pods.sort((a, b) => {
    return a.name.localeCompare(b.name)
  })

  return {
    data: pods,
    isLoading: !error && !data,
    error,
  }
}

export function usePod(podName: string): PodResponse {
  const path = getApiPath(`/pods/${podName}`)
  const { data, error } = useSWR(path, fetcher, {
    refreshInterval: 250,
  })
  const [pod, setPod] = useState<Pod>()

  useEffect(() => {
    if (data) {
      setPod(data)
    }
  }, [data])

  return {
    data: pod,
    isLoading: !error && !data,
    error,
  }
}
